pytensor/graph/rewriting/nested.py

```diff
@@ -0,0 +1,135 @@
+from typing import Optional
+
+from pytensor.compile.builders import OpFromGraph
+from pytensor.graph.basic import Apply, io_toposort
+from pytensor.graph.fg import FunctionGraph
+from pytensor.graph.replace import clone_replace
+from pytensor.graph.rewriting.basic import NodeRewriter, WalkingGraphRewriter
+from pytensor.scan.op import Scan
+
+
+class WalkingNestedGraphRewriter(WalkingGraphRewriter):
+    def process_node(
+        self,
+        fgraph: FunctionGraph,
+        node: Apply,
+        node_rewriter: Optional[NodeRewriter] = None,
+    ):
+        self.node_rewriter = node_rewriter or self.node_rewriter
+        if isinstance(node.op, (Scan, OpFromGraph)):
+            return self.process_scan_node(fgraph, node)
+        else:
+            return super().process_node(fgraph, node)
+
+    def process_scan_node(self, fgraph: FunctionGraph, node: Apply):
+        try:
+            replacements = self.transform_scan_node(fgraph, node)
+        except Exception as e:
+            if self.failure_callback is not None:
+                self.failure_callback(
+                    e,
+                    self,
+                    [(x, None) for x in node.outputs],
+                    self.node_rewriter,  # type: ignore
+                    node,
+                )
+                return False
+            else:
+                raise
+        if replacements is False or replacements is None:
+            return False
+
+        repl_pairs = zip(node.outputs, replacements)
+        try:
+            fgraph.replace_all_validate_remove(  # type: ignore
+                repl_pairs,
+                reason=self.node_rewriter,
+                remove=[],
+            )
+            return True
+        except Exception as e:
+            # This means the replacements were rejected by the fgraph.
+            #
+            # This is not supposed to happen.  The default failure_callback
+            # will print a traceback as a warning.
+            if self.failure_callback is not None:
+                self.failure_callback(
+                    e,
+                    self,
+                    repl_pairs,  # type: ignore
+                    self.node_rewriter,  # type: ignore
+                    node,
+                )
+                return False
+            else:
+                raise
+
+    def transform_scan_node(self, fgraph, node):
```

Comment:
I think we should try a simpler approach. A WalkingGraphRewriter already has all the logic to apply rewrites over a graph. Can we just use it in the inner graph of the Op?

Perhaps inside `process_node` we have something like:
```python
if isinstance(node.op, HasInnerGraphOp):
  self.apply(node.op.fgraph)
```